### PR TITLE
narinfo: more validation + underlying data structures

### PR DESCRIPTION
**Pull request by flokli**
This teaches go-nix how to parse signature lines and checksums, and provides the parsed version them it in the NarInfo.

For this to work, the `hash.Hash` and `nar.narinfo.Signature` structs, surrounding functions and corresponding tests were introduced.

There's also a `path` package now, which provides some basic sanity-checking for store paths.

Together with the nar.narinfo.NarInfo.Validate()` function, this provides some pretty good coverage in checking for invalid `.narinfo` files.

### CODE REVIEW

File: hash/hash.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+// hashFunc returns the cryptographic hash function for the passed HashType (implementing crypto.Hash)
+func hashFunc(hashType HashType) crypto.Hash {
+	switch hashType {
+	case HashTypeSha256:
+		return crypto.SHA256
+	case HashTypeSha512:
+		return crypto.SHA512
+	default:
+		panic(fmt.Sprintf("Invalid hash type: %v", hashType))
```
Comment (zimbatm) [inline hash/hash.go:31]: return error

File: hash/hash.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	}, nil
+}
+
+// String returns the string representation of a given hash
+// This is the hash type, a colon, and then the nixbase32-encoded digest
+// If the hash is inconsistent (digest size doesn't match hash type, an empty
+// string is returned)
+func (h *Hash) String() string {
+	if hashFunc(h.HashType).Size() != len(h.Digest) {
+		return ""
```
Comment (zimbatm) [inline hash/hash.go:81]: `panic("BUG: logic error")`

File: hash/hash.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		return crypto.SHA256
+	case HashTypeSha512:
+		return crypto.SHA512
+	default:
+		panic(fmt.Sprintf("Invalid hash type: %v", hashType))
+	}
+}
+
+// NewHash returns a new Hash struct, by parsing a hashtype:nixbase32 string
+func NewHashFromNixBase32(s string) (*Hash, error) {
```
Comment (zimbatm) [inline hash/hash.go:36]:
```suggestion
// ParseNixBase32 returns a new Hash struct, by parsing a hashtype:nixbase32 string
func ParseNixBase32(s string) (*Hash, error) {
```

File: nar/narinfo/validate.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	"github.com/google/go-cmp/cmp"
+	nixpath "github.com/numtide/go-nix/path"
+	"path"
+)
+
+// Validate does some sanity checking on a NarInfo struct, such as:
+// - ensuring the paths in StorePath, References and Deriver are syntactically valid
+//   (references and deriver first need to be made absolute)
+// - when no compression is present, ensuring File{Hash,Size} and Nar{Hash,Size} are equal
+func (n *NarInfo) Validate() error {
```
Comment (zimbatm) [inline nar/narinfo/validate.go:14]:
```suggestion
// Check does some sanity checking on a NarInfo struct, such as:
// - ensuring the paths in StorePath, References and Deriver are syntactically valid
//   (references and deriver first need to be made absolute)
// - when no compression is present, ensuring File{Hash,Size} and Nar{Hash,Size} are equal
func (n *NarInfo) Check() error {
```

File: path/nixpath.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+func (n *NixPath) String() string {
+	return path.Join(StoreDir, fmt.Sprintf("%v-%v", nixbase32.EncodeToString(n.Digest), n.Name))
+}
+
+// NewFromString parses a path string into a nix path,
+// verifying it's syntactically valid
+// It returns an error if it fails to parse
+func NewFromString(s string) (*NixPath, error) {
```
Comment (zimbatm) [inline path/nixpath.go:39]:
```suggestion
// FromString parses a path string into a nix path,
// verifying it's syntactically valid
// It returns an error if it fails to parse
func FromString(s string) (*NixPath, error) {
```

File: path/nixpath.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
+package path
```
Comment (zimbatm) [inline path/nixpath.go:1]:
```suggestion
package nixpath
```

**Review by zimbatm — APPROVED** [review]

**Comment by flokli** [comment]
Thanks for the review! I addressed all the remarks.

**Review by zimbatm — APPROVED** [review]

**Pull request merged**